### DISCUSSION TITLE
Dynamic Route Configuration At Runtime

### DISCUSSION THREAD

**Discussion by jakehamilton** [fastify/fastify#3134] [Ideas]
Hey there folks! I've been using Fastify for a bit now and have really appreciated a lot of the API decisions made. It's ended up fitting more cleanly into my projects than Express which is pretty impressive. However, I recently ran into an issue where I needed to:

1. Create a fastify instance
2. Setup routes
3. Listen on a port
4. Reconfigure routes

The current version of Fastify doesn't support this, throwing an error:

```
Error: Cannot add route when fastify instance is already started!
```

I'd done some reading previously and on other threads the consensus seemed to be that the routing library [`find-my-way`](https://www.npmjs.com/package/find-my-way) didn't support removing routes. From a cursory glance it seems that `find-my-way` currently supports [`router.off`](https://www.npmjs.com/package/find-my-way#offmethods-path-handler-store). Perhaps this is something that could be integrated?

I know that the vast majority of applications using Fastify right now likely won't find this beneficial. But I do think that this could open the door for some very useful tooling. Currently, to accomplish the same effect, I'd have to totally side-step Fastify and its handlers in order to make something this dynamic work. Or worse, completely destroy and recreate the server whenever a route is changed.

If anyone has any thoughts on this, I'd love to hear them!

**Comment by zekth** [comment]
Can you describe one case where this could be better than using parametrized urls?

**Reply by jakehamilton** [reply]
Hey @zekth!

The most immediate problem I'm working on is configuring arbitrary routes at runtime with support for middleware and all of the other handlers/configuration that Fastify provides per-route for rapid development. I don't want to go too deep into the specifics, but can explain further if you'd like. That said:

I have a system that breaks up different pieces like routes into "units" that provide something like this:

```js
{
  config: {
    path: "/",
    method: "GET"
  },
  handler: (request, reply) => {},
  // other handlers for `fastify.route()`
}
```

These are all loaded on startup and applied to the Fastify instance that is created. This works great if done before `fastify.listen()` is called. But one important feature in this system is the ability to **uninstall** units. In this case, that would mean removing the route from the Fastify instance. One use for this feature is to hot-swap units during development to enable a similar experience to that of modern React with `react-refresh`. It is also quite useful to be able to dynamically configure some things at runtime in production, but this doesn't happen often for routes.

Given these requirements, I don't think parameterized URLs would solve the problem since I'd have to reimplement the routing, handlers, and middleware layer myself. Additionally, it'd be a bit cumbersome to allow for Fastify plugins to interact nicely.

---

In case more context is useful, the project that I'm working on is [Leverage](https://github.com/jakehamilton/leverage/tree/main/packages/core) and its [HTTP Plugin](https://github.com/jakehamilton/leverage/tree/main/packages/plugin-http). I have hacked up a working version which requires destroying the Fastify server whenever a route is changed after the server starts listening. Here's the [important bit](https://github.com/jakehamilton/leverage/blob/feat/dynamic-routes/packages/plugin-http/src/plugin.js#L295).
